### PR TITLE
UBX-NAV-TIMELS support added for leap second event

### PR DESCRIPTION
**Pull request by UT2UH**
Provides leap second event info as SNTP Leap Indicator, time to a leap second event and the number of leap seconds since GPS epoch for applications like NTP server.

### CODE REVIEW

**Comment by PaulZC** [comment]
Hi @UT2UH ,
This PR looks really nice - thank you!
The only thing that jumps out is that the library has a new ```end``` function which deletes all of the RAM allocated for message storage. We will need to include the TIMELS storage in there too. You are not using callbacks - which is OK - so the code will look the same as [moduleSWVersion and currentGeofenceParams](https://github.com/sparkfun/SparkFun_u-blox_GNSS_Arduino_Library/blob/master/src/SparkFun_u-blox_GNSS_Arduino_Library.cpp#L74-L84). Can you please add it? Thanks!
Best wishes,
Paul

**Comment by UT2UH** [comment]
Hello @PaulZC , 
hope that is it. 
Thanks and regards!

File: src/SparkFun_u-blox_GNSS_Arduino_Library.cpp [hunk 1]
```diff
@@ -10304,6 +10383,45 @@ float SFE_UBLOX_GNSS::getSurveyInMeanAccuracy(uint16_t maxWait) // Returned as m
...
+  return ((boolean)packetUBXNAVTIMELS->data.valid ? (uint8_t)(packetUBXNAVTIMELS->data.lsChange == -1 ? 2 : packetUBXNAVTIMELS->data.lsChange) : 3);
+}
+
+int8_t SFE_UBLOX_GNSS::getCurrentLeapSeconds(sfe_ublox_ls_src_e& source, uint16_t maxWait)
+{
+  if (packetUBXNAVTIMELS == NULL) initPacketUBXNAVTIMELS(); //Check that RAM has been allocated for the TIMELS data
+  if (packetUBXNAVTIMELS == NULL) //Bail if the RAM allocation failed
+    return false;
+
+  if (packetUBXNAVTIMELS->moduleQueried.moduleQueried.bits.valid == false)
```
Comment (PaulZC) [inline src/SparkFun_u-blox_GNSS_Arduino_Library.cpp:10415]: Hi @UT2UH ,
If you use ```.valid``` here to check if the currLs data has been read before, then if you call ```getLeapIndicator``` followed by ```getCurrentLeapSeconds``` then ```getCurrentLeapSeconds``` will always request fresh data (by calling ```getLeapSecondEvent```). I'm not sure if that is what you want? I suspect probably not?
My advice would be to change ```.valid``` in line 10415 to ```.currLs``` and then delete line 10417.
That way ```getLeapIndicator``` will only be called once if you call ```getLeapIndicator``` followed by ```getCurrentLeapSeconds``` or vice versa.
Please have a think about this. Thanks!
Paul

File: src/u-blox_structs.h [hunk 2]
```diff
@@ -854,6 +854,57 @@ typedef struct
...
+  uint8_t reserved1[3];
+  uint8_t srcOfCurrLs; //Information source for the current number of leap seconds
+  int8_t currLs; //Current number of leap seconds since start of GPS (Jan 6, 1980), s
+  uint8_t srcOfLsChange; //Information source for the future leap second event
+  int8_t lsChange; //Future leap second change if one is scheduled, +1, 0, -1s
+  int32_t timeToLsEvent; //Num of secs until the next or from the last leap second, s
+  uint16_t dateOfLsGpsWn; //GPS week num (WN) of the next or the last leap second event
+  uint16_t dateOfLsGpsDn; //GPS day of week num (DN) for the next or last leap second event
+  uint8_t reserved2[3];
+  int8_t valid; // Validity flag, 1 = valid, otherwise 0
```
Comment (PaulZC) [inline src/u-blox_structs.h:873]: Hi @UT2UH ,
Looking at the ZED_F9P interface description, ```valid``` contains two bits (not one):

image

(Sometimes the ZED messages contain extra bits compared to the M8 family)
Can you please change ```valid``` into a union+struct? Something like:

```
  union
  {
    uint8_t all;
    struct
    {
      uint8_t validCurrLs : 1; // 1 = Valid current number of leap seconds value
      uint8_t validTimeToLsEvent : 1; // 1 = Valid time to next leap second event or from the last leap second event if no future event scheduled
    } bits;
  } valid;
```

This means you will need to change ```getLeapIndicator``` so it uses ```data.valid.bits.validCurrLs```.
Thanks!
Paul

File: src/u-blox_structs.h [hunk 3]
```diff
@@ -854,6 +854,57 @@ typedef struct
...
+      uint32_t iTOW : 1;
+      uint32_t version : 1;
+      uint32_t srcOfCurrLs : 1;
+      uint32_t currLs : 1;
+      uint32_t srcOfLsChange : 1;
+      uint32_t lsChange : 1;
+      uint32_t timeToLsEvent : 1;
+      uint32_t dateOfLsGpsWn : 1;
+      uint32_t dateOfLsGpsDn : 1;
+      uint32_t valid : 1;
```
Comment (PaulZC) [inline src/u-blox_structs.h:894]: Sorry - this needs to be:
```
uint32_t validCurrLs : 1;
uint32_t validTimeToLsEvent : 1;
```

**Review by PaulZC — CHANGES_REQUESTED** [review]
Hi @UT2UH ,
This is a really nice PR - thank you. You have followed the structure of the "auto" message code really well. I just have three comments (see above). Please take a look...
Thanks again,
Paul

**Comment by UT2UH** [comment]
Hi @PaulZC !
Yep, you are right, including on my mindless copycat issues too... Strange that it worked somehow... )
Thank you very much for the help and for the great library!

**Comment by PaulZC** [comment]
Thanks @UT2UH ,
Good use of those two valid bits! ;-)
Merging...
Thanks again,
Paul

**Pull request merged**